### PR TITLE
Add proper func tests

### LINKED ISSUE #5: Add a test CI workflow
We should have a workflow to run against PRs for common checks/tests/lints:

- shellcheck the hook scripts
- yaml lint the snapcraft.yaml
- verify clean build with snapcraft

### PR DESCRIPTION
**Pull request by Deezzir**
- Adds the functional tests
- Adds the check workflow https://github.com/canonical/solutions-engineering-automation/issues/122
- Adds the tics workflow https://github.com/canonical/solutions-engineering-automation/issues/115

Closes: #5

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -126,6 +126,14 @@ adopt-info: openstack-exporter
 confinement: strict
 grade: stable
 
+platforms:
+  amd64:
+    build-on: [amd64]
+    build-for: [amd64]
+  arm64:
+    build-on: [arm64]
+    build-for: [arm64]
```
Comment (samuelallan72) [inline snap/snapcraft.yaml]:
```suggestion
  amd64:
  arm64:
```

Suggesting this to make it easier to read with less duplication. See https://snapcraft.io/docs/reference-architectures :

>  If the platform name is a valid debian architecture, then build-on and build-for can be omitted (see below for details).

File: tests/functional/conftest.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import subprocess
+
+import pytest
+
+
+@pytest.fixture(scope="session", autouse=True)
+def install_dcgm_snap():
+    """Install the snap and enable openstack-exporter service for testing."""
+    snap_build_name = "charmed-openstack-exporter_*.snap"
```
Comment (samuelallan72) [inline tests/functional/conftest.py:9]: I like this method, and nice that it replaces an otherwise required environment variable in the tests. If you have time, could you check our other snaps where we use the TEST_SNAP variable, and see where we can update the tests to this new pattern? Then we can remove the TEST_SNAP variable from the check.yaml template. :+1:

File: tests/functional/conftest.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        ]
+    )
+    subprocess.check_call(["sudo", "snap", "start", "charmed-openstack-exporter"])
+
+    yield
+
+    subprocess.check_call(["sudo", "snap", "remove", "--purge", "charmed-openstack-exporter"])
+    subprocess.check_call(["sudo", "rm", config_file_path])
+    subprocess.check_call(["sudo", "rm", config_test_file_path])
+    subprocess.check_call(["sudo", "rmdir", config_dir_path])
```
Comment (samuelallan72) [inline tests/functional/conftest.py:43]: Because of all these destructive operations in the func tests, could you add a warning to the readme and tox.ini target (and anywhere else appropriate?) about running it locally on your workstation?

**Review by samuelallan72 — CHANGES_REQUESTED** [review]
Thanks, this looks great in general! I left a request about warning users about running func tests locally due to the destructive actions.

(I think a warning like this is something we should add to other projects too - wherever they have destructive local actions. It's not necessarily expected that func tests will mess with your local system - eg. most charm func tests are safe to run locally because they only interact with an ephemeral juju model).

File: tests/functional/conftest.py (hunk 2 above)
Comment (Deezzir) [reply tests/functional/conftest.py:9]: I just noticed the name of the function :)

File: tests/functional/conftest.py (hunk 3 above)
Comment (Deezzir) [reply tests/functional/conftest.py:43]: Added a warning to the README file

File: tests/functional/conftest.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    config_dir_path = "/etc/openstack"
+    config_file_path = f"{config_dir_path}/clouds.yaml"
+    config_test_file_path = f"{config_dir_path}/test.yaml"
+
+    subprocess.check_call(
+        f"sudo snap install --dangerous {snap_build_name}",
+        shell=True,
+    )
+
+    # Create the directory if it doesn't exist and create the file
```
Comment (aieri) [inline tests/functional/conftest.py]: non-blocking: which file?

File: tests/functional/conftest.py (hunk 3 above)
Comment (aieri) [reply tests/functional/conftest.py:43]: could we store the config file in TMPDIR instead? It'd still be nice to clean up after ourselves, but risks would be minimized

File: tests/functional/test_snap.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    status_code = response.getcode()
+    assert status_code == 200, f"Endpoint {endpoint} returned status code {status_code}"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _assert_service_listening(bind: str, service: str) -> None:
+    """Assert service is listening on a specific bind."""
+    pid = subprocess.check_output(["sudo", "lsof", "-t", "-i", bind], text=True).strip()
+    assert service in subprocess.check_output(
+        ["cat", f"/proc/{pid}/cmdline"], text=True
```
Comment (aieri) [inline tests/functional/test_snap.py]: suggestion: avoid UUOC by using native file reading

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 name: charmed-openstack-exporter
-base: core22
+base: core24
```
Comment (aieri) [inline snap/snapcraft.yaml]: question: do the tests depend on using the new core? Could we reserve this change for a separate PR instead?

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -126,6 +126,10 @@ adopt-info: openstack-exporter
 confinement: strict
 grade: stable
 
+platforms:
+  amd64:
+  arm64:
+
```
Comment (aieri) [inline snap/snapcraft.yaml]: question: could the introduction of arm64 be reserved for a separate PR?

File: tests/functional/conftest.py (hunk 3 above)
Comment (Deezzir) [reply tests/functional/conftest.py:43]: We can't its a confined snap and only has access to that specific dir provided through the plug

Comment (Deezzir) [reply tests/functional/conftest.py:43]: https://github.com/canonical/charmed-openstack-exporter-snap/blob/6096b2fda9b29ad6285a2f7bc0c387c80f59efb8/snap/snapcraft.yaml#L162

When I saw that, I realized we could do the same for the dcgm-snap, avoiding the need for the user to put the files in the snap's folder - speaking about UX. But I didn't know that we can do this :(

File: snap/snapcraft.yaml (hunk 6 above)
Comment (Deezzir) [reply snap/snapcraft.yaml]: No, you're right, this should be a separate PR along with the platforms stuff

File: tests/functional/conftest.py (hunk 3 above)
Comment (samuelallan72) [reply tests/functional/conftest.py:43]: Don't worry too much - putting files in the snap folder is preferrable if possible. The snap needs a separate review process for any system-files interface too, and would probably only be approved if it's strictly necessary. :)

File: tests/functional/conftest.py (hunk 2 above)
Comment (samuelallan72) [reply tests/functional/conftest.py:9]: Oh aha I didn't even see that :+1:

File: .github/workflows/check.yaml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+          yamllint .yamllint snap/snapcraft.yaml
+
+  build:
+    needs:
+      - lint
+    runs-on: ${{ matrix.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        runs-on: [[ubuntu-22.04], [self-hosted, jammy, ARM64]]
```
Comment (samuelallan72) [inline .github/workflows/check.yaml]: If you're adding arm64 to snapcraft.yaml in a later PR, you'll probably want to remove the arm64 runs-on here.

**Review by samuelallan72 — APPROVED** [review]

File: tests/functional/conftest.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+@pytest.fixture(scope="session", autouse=True)
+def install_charmed_openstack_exporter_snap():
+    """Install the snap and enable openstack-exporter service for testing."""
+    snap_build_name = "charmed-openstack-exporter_*.snap"
+    config_dir_path = "/etc/openstack"
+    config_file_path = f"{config_dir_path}/clouds.yaml"
+    config_test_file_path = f"{config_dir_path}/test.yaml"
+
+    subprocess.check_call(
```
Comment (jneo8) [inline tests/functional/conftest.py:14]: **suggestion (non-blocking):** 

What if the snap file not exists, can we raise an error here?

File: tests/functional/conftest.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        ]
+    )
+    subprocess.check_call(["sudo", "snap", "start", "charmed-openstack-exporter"])
+
+    yield
+
+    subprocess.check_call(["sudo", "snap", "remove", "--purge", "charmed-openstack-exporter"])
+    subprocess.check_call(["sudo", "rm", config_file_path])
+    subprocess.check_call(["sudo", "rm", config_test_file_path])
+    subprocess.check_call(["sudo", "rmdir", config_dir_path])
```
Comment (jneo8) [inline tests/functional/conftest.py:43]: **suggestion (non-blocking):** 

There is a possibility of side effects. For instance, if you create a file but the `sudo snap set ...` command fails, the created files will not be automatically removed.

**Review by jneo8 — APPROVED** [review]
Two suggestions comments about defensive programming. These are optional improvements, so feel free to address them as you see fit.

Other part is LGTM.
